clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -833,7 +838,7 @@ public void testFetchPositionAfterException() {
         partitions.put(tp1, new FetchResponse.PartitionData(Errors.NONE, 100,
             FetchResponse.INVALID_LAST_STABLE_OFFSET, FetchResponse.INVALID_LOG_START_OFFSET, null, records));
         partitions.put(tp0, new FetchResponse.PartitionData(Errors.OFFSET_OUT_OF_RANGE, 100,
-                FetchResponse.INVALID_LAST_STABLE_OFFSET, FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY));
+            FetchResponse.INVALID_LAST_STABLE_OFFSET, FetchResponse.INVALID_LOG_START_OFFSET, null, outOfOrderRecords));
```

Comment:
In reality when OffsetOutOfRangeException is thrown, the records will be empty. In this case, if the auto offset reset is not enabled, an exception should be thrown to the users. The current patch will miss this exception. This is because the exception will be caught by the outer loop in `Fetcher.fetchRecords()` and ignored when there are fetched records. The next `Fetcher.fetchRecords()` call will not see any exception because the `CompletedFetch` with the error code has been polled out of the queue. It looks that we should not poll the `CompletedFetch` out of the queue if the `fetched` is not empty.